File: webapp/src/lib/generate-avatar.ts
import { supabase } from './supabase'
import type { AvatarConfig } from './types'

const FN_URL = `${import.meta.env.VITE_SUPABASE_URL}/functions/v1/generate-avatar`

/**
 * Premium: ask the `generate-avatar` edge function to analyze a member's photo
 * with Claude vision and return a matching DiceBear avatar config. Server-side
 * (the Anthropic key never touches the client).
 */
export async function generateAvatarFromPhoto(memberId: string): Promise<AvatarConfig> {
  const { data: sess } = await supabase.auth.getSession()
  const jwt = sess.session?.access_token
  if (!jwt) throw new Error('Not signed in')
  const res = await fetch(FN_URL, {
    method: 'POST',
    headers: { Authorization: `Bearer ${jwt}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ memberId }),
  })
  if (!res.ok) {
    const t = await res.text().catch(() => '')
    throw new Error(`Generation failed: ${res.status} ${t}`)
  }
  return (await res.json()) as AvatarConfig
}

File: lib/src/features/avatars/data/generate_avatar.dart
import 'dart:convert';
import 'dart:io';

import 'package:supabase_flutter/supabase_flutter.dart';

import '../../../config/app_config.dart';

/// Premium: call the `generate-avatar` edge function to analyze a member's photo
/// with Claude vision and return a matching DiceBear avatar config map. The
/// Anthropic key lives server-side; this only sends the user's JWT + memberId.
Future<Map<String, dynamic>> generateAvatarFromPhoto(
    SupabaseClient client, String memberId) async {
  final jwt = client.auth.currentSession?.accessToken;
  if (jwt == null) throw Exception('Not signed in');
  final url = '${AppConfig.supabaseUrl}/functions/v1/generate-avatar';

  final http = HttpClient();
  try {
    final req = await http.postUrl(Uri.parse(url));
    req.headers.set(HttpHeaders.authorizationHeader, 'Bearer $jwt');
    req.headers.set(HttpHeaders.contentTypeHeader, 'application/json');
    req.add(utf8.encode(jsonEncode({'memberId': memberId})));
    final resp = await req.close();
    final body = await resp.transform(utf8.decoder).join();
    if (resp.statusCode != 200) {
      throw Exception('Generation failed: ${resp.statusCode} $body');
    }
    return jsonDecode(body) as Map<String, dynamic>;
  } finally {
    http.close();
  }
}
